tracing/trace.go

```diff
@@ -0,0 +1,272 @@
+// Using the information in the golang stack-trace,
+//a context.Context environment is created and then passed to the tracing
+//library.
+//
+// Currently there are two implementations of the Trace and Span:
+//  - simul for tests that collect all traces in a simulCollector
+//  - honeycomb that sends the traces to honeycomb
+// It should be possible to add other tracing modules,
+//like prometheus.io or zipkin.io.
+// The implementation of these tracers is left as an exercise to the reader...
+
+// Trace interfaces the methods that are used by the wrappers,
+// for easier testing and eventual usage of other tools.
+type Trace interface {
+	// AddFiled adds a key/val to the trace.
+	//This key/val pair will be set to all spans contained in this trace.
+	AddField(key string, val interface{})
+	// There is at least one span for each trace, which is the root span.
+	GetRootSpan() Span
+	// Finishes the trace and sends it to the main service.
+	// Once Send is called,
+	//further interactions with the Trace structure are not defined.
+	Send()
+}
+
+// Span interfaces the methods that are used by the wrappers,
+// for easier testing and eventual usage of other tools than HoneyComb.
+type Span interface {
+	AddField(key string, val interface{})
+	CreateChild(ctx context.Context) (context.Context, Span)
+	Send()
+}
+
+// traceWrapper keeps track of the stackEntries and creates / sends Spans as
+// needed.
+type traceWrapper struct {
+	ctx     context.Context
+	root    *spanWrapper
+	hcTrace Trace
+}
+
+func newTraceWrapper(ctx context.Context, tr Trace,
+	ses []stackEntry) (tw *traceWrapper, child *spanWrapper) {
+	tw = &traceWrapper{
+		ctx:     ctx,
+		hcTrace: tr,
+	}
+	tw.root, child = newRootSpanWrapper(tw, ses)
+	return
+}
+
+func (tr *traceWrapper) stackToSpan(stack []stackEntry) *spanWrapper {
+	return tr.root.stackToSpan(stack)
+}
+
+func (tr *traceWrapper) send() {
+	tr.hcTrace.Send()
+}
+
+// spanWrapper keeps track of the contexts created for the different spans
+// and closes children as new spans arrive.
+type spanWrapper struct {
+	se     stackEntry
+	ctx    context.Context
+	trace  *traceWrapper
+	parent *spanWrapper
+	child  *spanWrapper
+	hcSpan Span
+}
+
+func newSpanWrapper(ctx context.Context, t *traceWrapper, hcs Span, se stackEntry) *spanWrapper {
+	newSpan := &spanWrapper{
+		ctx:    ctx,
+		trace:  t,
+		hcSpan: hcs,
+		se:     se,
+	}
+	se.addFields(hcs)
+	return newSpan
+}
+
+func newRootSpanWrapper(t *traceWrapper, stack []stackEntry) (root,
+	child *spanWrapper) {
+	root = newSpanWrapper(t.ctx, t, t.hcTrace.GetRootSpan(), stack[0])
+	if len(stack) > 1 {
+		child = root.stackToSpan(stack)
+	} else {
+		child = root
+	}
+	return
+}
+
+func (sp *spanWrapper) log(lvl int, msg string) {
+	if sp.child != nil {
+		sp.child.done()
+	}
+	_, sLog := sp.hcSpan.CreateChild(sp.ctx)
+	msgStruct := struct {
+		Lvl int
+		Msg string
+	}{lvl, msg}
+	// This value is unfortunately not exported from onet/log/lvl.go
+	if lvl > -17 {
```

Comment:
Magic?

Reply:
Yes, explained in the comment above...

Reply:
It would be helpful to document it refers to `lvlPanic` though. :)